### PR TITLE
Fix case-sensitivity failure for DISTINCT + ORDER BY

### LINKED ISSUE #4225: Qualified distinct columns are case sensitive
```sql
SELECT DISTINCT o.CUSTKEY FROM orders o ORDER BY o.custkey LIMIT 5;
```
```
io.prestosql.spi.PrestoException: line 1:1: For SELECT DISTINCT, ORDER BY expressions must appear in select list
	at io.prestosql.sql.analyzer.SemanticExceptions.semanticException(SemanticExceptions.java:48)
	at io.prestosql.sql.analyzer.SemanticExceptions.semanticException(SemanticExceptions.java:43)
	at io.prestosql.sql.analyzer.StatementAnalyzer$Visitor.verifySelectDistinct(StatementAnalyzer.java:2651)
	at io.prestosql.sql.analyzer.StatementAnalyzer$Visitor.visitQuerySpecification(StatementAnalyzer.java:1303)
	at io.prestosql.sql.analyzer.StatementAnalyzer$Visitor.visitQuerySpecification(StatementAnalyzer.java:306)
	at io.prestosql.sql.tree.QuerySpecification.accept(QuerySpecification.java:144)
	at io.prestosql.sql.tree.AstVisitor.process(AstVisitor.java:27)
	at io.prestosql.sql.analyzer.StatementAnalyzer$Visitor.process(StatementAnalyzer.java:323)
	at io.prestosql.sql.analyzer.StatementAnalyzer$Visitor.process(StatementAnalyzer.java:333)
	at io.prestosql.sql.analyzer.StatementAnalyzer$Visitor.visitQuery(StatementAnalyzer.java:918)
	at io.prestosql.sql.analyzer.StatementAnalyzer$Visitor.visitQuery(StatementAnalyzer.java:306)
	at io.prestosql.sql.tree.Query.accept(Query.java:107)
	at io.prestosql.sql.tree.AstVisitor.process(AstVisitor.java:27)
	at io.prestosql.sql.analyzer.StatementAnalyzer$Visitor.process(StatementAnalyzer.java:323)
	at io.prestosql.sql.analyzer.StatementAnalyzer.analyze(StatementAnalyzer.java:292)
	at io.prestosql.sql.analyzer.Analyzer.analyze(Analyzer.java:83)
	at io.prestosql.sql.analyzer.Analyzer.analyze(Analyzer.java:75)
	at io.prestosql.execution.SqlQueryExecution.analyze(SqlQueryExecution.java:235)
	at io.prestosql.execution.SqlQueryExecution.<init>(SqlQueryExecution.java:181)
	at io.prestosql.execution.SqlQueryExecution$SqlQueryExecutionFactory.createQueryExecution(SqlQueryExecution.java:732)
	at io.prestosql.dispatcher.LocalDispatchQueryFactory.lambda$createDispatchQuery$0(LocalDispatchQueryFactory.java:121)
	at io.prestosql.$gen.Presto_testversion____20200625_152010_4.call(Unknown Source)
	at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:125)
	at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:69)
	at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:78)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

### PR DESCRIPTION
**Pull request by martint**
Fixes https://github.com/prestosql/presto/issues/4225

### CODE REVIEW

File: presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java [hunk 1]
```diff
@@ -2597,21 +2545,56 @@ private Scope analyzeWith(Query node, Optional<Scope> scope)
             return withScope;
         }
 
-        private void verifySelectDistinct(QuerySpecification node, List<Expression> outputExpressions)
+        private Set<CanonicalizationAware<Identifier>> getAliases(Select node)
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java:2548]: move this method below `verifySelectDistinct`

**Review by kokosing — APPROVED** [review]

File: presto-main/src/test/java/io/prestosql/sql/query/TestDistinctWithOrderBy.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+    @AfterClass(alwaysRun = true)
+    public void teardown()
+    {
+        assertions.close();
+        assertions = null;
+    }
+
+    @Test
+    public void testOrderByReferenceWithDifferentCapitalization()
```
Comment (kasiafi) [inline presto-main/src/test/java/io/prestosql/sql/query/TestDistinctWithOrderBy.java:40]: it is `withDifferentQualification`, too

File: presto-main/src/test/java/io/prestosql/sql/query/TestDistinctWithOrderBy.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+                .matches("VALUES 2, 4");
+    }
+
+    @Test
+    public void testSelectAllAliases()
+    {
+        assertThat(assertions.query("SELECT DISTINCT t.r.* AS (a, b) FROM (VALUES ROW(CAST(ROW(1,1) AS ROW(a BIGINT, b BIGINT)))) t(r) ORDER BY a"))
+                .matches("VALUES (BIGINT '1', BIGINT '1')");
+    }
+}
```
Comment (kasiafi) [inline presto-main/src/test/java/io/prestosql/sql/query/TestDistinctWithOrderBy.java:73]: Consider adding a test with conflicting aliases, e.g. `SELECT DISTINCT a, b a FROM (VALUES (1, 5), (2, 4)) T(a,b) ORDER BY T.a DESC`

File: presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java [hunk 4]
```diff
@@ -2597,21 +2545,56 @@ private Scope analyzeWith(Query node, Optional<Scope> scope)
...
+                    Optional<Identifier> alias = column.getAlias();
+                    if (alias.isPresent()) {
+                        aliases.add(canonicalizationAwareKey(alias.get()));
+                    }
+                    else if (column.getExpression() instanceof Identifier) {
+                        aliases.add(canonicalizationAwareKey((Identifier) column.getExpression()));
+                    }
+                    else if (column.getExpression() instanceof DereferenceExpression) {
+                        aliases.add(canonicalizationAwareKey(((DereferenceExpression) column.getExpression()).getField()));
+                    }
```
[the remark this answers is not in the document]
Comment (martint) [reply presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java:2563]: Exactly, if you have `SELECT a FROM t ORDER BY a`, the `a` in the `SELECT` clause is bound to the `FROM` scope, while the `a` in `ORDER BY` clause is bound to the `a` from the `SELECT` clause, so we can't compare by field id / relation id.

**Pull request merged**